src/Compilers/Core/RebuildTest/DeterministicKeyBuilderTests.cs

```diff
@@ -0,0 +1,703 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.Emit;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.VisualBasic;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Roslyn.Test.Utilities;
+using Roslyn.Test.Utilities.TestGenerators;
+using Roslyn.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Rebuild.UnitTests
+{
+    public abstract class DeterministicKeyBuilderTests<TCompilation, TCompilationOptions, TParseOptions>
+        where TCompilation : Compilation
+        where TCompilationOptions : CompilationOptions
+        where TParseOptions : ParseOptions
+    {
+        private static readonly char[] s_trimChars = { ' ', '\n', '\r' };
+
+        public static SourceHashAlgorithm HashAlgorithm { get; } = SourceHashAlgorithm.Sha256;
+        public static SourceHashAlgorithm[] HashAlgorithms { get; } = new[]
+        {
+            SourceHashAlgorithm.Sha1,
+            SourceHashAlgorithm.Sha256
+        };
+
+        protected static void AssertJson(
+            string expected,
+            string actual,
+            bool removeStandard = true) => AssertJson(expected, actual, "references", "extensions");
+
+        protected static void AssertJson(
+            string expected,
+            string actual,
+            params string[] ignoreSections)
+        {
+            var json = JObject.Parse(actual);
+            if (ignoreSections.Length > 0)
+            {
+                json
+                    .Descendants()
+                    .OfType<JProperty>()
+                    .Where(x => ignoreSections.Contains(x.Name))
+                    .ToList()
+                    .ForEach(x => x.Remove());
+            }
+
+            actual = json.ToString(Formatting.Indented);
+            expected = JObject.Parse(expected).ToString(Formatting.Indented);
+            AssertJsonCore(expected, actual);
+        }
+
+        protected static void AssertJsonSection(
+            string expected,
+            string actual,
+            string sectionName,
+            params string[] ignoreProperties)
+        {
+            var property = GetJsonProperty(actual, sectionName, ignoreProperties);
+            AssertJsonCore(expected, property.ToString(Formatting.Indented));
+        }
+
+        protected static void AssertJsonCore(string? expected, string? actual)
+        {
+            expected = expected?.Trim(s_trimChars);
+            actual = actual?.Trim(s_trimChars);
+            Assert.Equal(expected, actual);
+        }
+
+        private protected static void AssertSyntaxTreePathMap(string? expected, CommonCompiler compiler)
```

Comment:
It wasn't clear to me what the utility was in making some methods `private protected`.

Reply:
`CommonCompiler` is `internal` so I had to chose between `internal` and `private protected` for these methods. The intent was to be consumed in derived types hence I chose `protected` route